test/arrayimpl.py

```diff
@@ -526,19 +512,23 @@ def as_raw(self):
             return self.array
 
 
-def subTest(case, skip=(), skiptypecode=()):
+def loop(*args):
     for array in ArrayBackends:
-        if array.backend == skip:
-            continue
-        if array.backend in skip:
-            continue
+        loop.array = array
         for typecode in array.TypeMap:
-            if typecode == skiptypecode:
-                continue
-            if typecode in skiptypecode:
-                continue
-            with case.subTest(backend=array.backend, typecode=typecode):
-                try:
-                    yield array, typecode
-                except GeneratorExit:
-                    return
+            loop.typecode = typecode
+            if not args:
+                yield array, typecode
+            else:
+                for prod in itertools.product(*args):
+                    yield (array, typecode) + prod
+    del loop.array
+    del loop.typecode
+
+
+def test(case, **kargs):
+    return case.subTest(
+        typecode=loop.typecode,
+        backend=loop.array.backend,
```

Comment:
and why would `loop.typecode` and `loop.array` be defined at this point?

Reply:
Because of the way `loop()` and `test()` are used. The point of these two routines, silly as they may look, it to reduce the verbosity of looping over arraytypes/typecodes and using the subTest context manager at the same time.